File: F_AI_Apocalypse/too_much_dark.cpp
#include<bits/stdc++.h>
using namespace std;
vector<vector<long long int>> adj;
vector<bool> visited;
vector<long long int> parent;

void dfs(long long int node) {
    visited[node] = true;
    for(long long int i = 0 ; i < adj[node].size() ; i++) {
        if(!visited[adj[node][i]]) {
            dfs(adj[node][i]);
            parent[adj[node][i]] = node;
        }
    }
}

long long int dfs2(long long int node , long long int x) {
    long long int ans = 0;
    for(long long int i = 0 ; i < adj[node].size() ; i++) {
        if(adj[node][i] == x) continue;
        ans = max(ans , dfs2(adj[node][i] , node));
    }
    return ans+1;
}

void solve() {
    long long int n , i , x , y;
    cin >> n;
    adj = vector<vector<long long int>> (n+1);
    visited = vector<bool> (n+1);
    parent = vector<long long int> (n+1 , -1);
    for(i = 1 ; i <= n-1 ; i++) {
        cin >> x >> y;
        adj[x].push_back(y);
        adj[y].push_back(x);
    }
    long long int p , q;
    cin >> x >> y;
    if(x == y) {
        cout << 0 << "\n";
        return;
    }
    p = x , q = y;
    vector<long long int> v1 , v2;
    map<long long int , long long int> mp;
    dfs(1);
    while(x != -1) {
        v1.push_back(x);
        x = parent[x];
    }
    while(y != -1) {
        mp[y] = 1;
        y = parent[y];
    }
    x = p , y = q;
    long long int lca;
    for(i = 0 ; i < v1.size() ; i++) {
        if(mp[v1[i]] == 1) {
            lca = v1[i];
            break;
        }
    }
    vector<long long int> v , temp;
    x = p , y = q;
    while(x != lca) {
        v.push_back(x);
        x = parent[x];
    }
    v.push_back(x);
    while(y != lca) {
        temp.push_back(y);
        y = parent[y];
    }
    reverse(temp.begin() , temp.end());
    for(i = 0 ; i < temp.size() ; i++) v.push_back(temp[i]);
    long long int mid = (v.size()-1)/2;
    long long int ans = dfs2(v[mid+1] , v[mid]);
    cout << ans+mid << "\n";
}

int main() {
    long long int tc;
    cin >> tc;
    while(tc-- > 0) {
        solve();
    }

}

File: E_Too_Much_Dark/Too_Much_Dark.java
/*
    author: KRHero
    IDE: VSCode
*/
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.StringTokenizer;
 

public class Too_Much_Dark {
    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        PrintWriter out = new PrintWriter(outputStream);
        Tasker task = new Tasker();
        long t = in.nextInt();
        while(t-->0){
            task.solve(in,out);
        }
        out.close();
    }

    static class Tasker{

        public int[] level;
        public int[] marthaThere;
        public int[] parents;
        int martha = 0;
        int[][] edges; 
        int[] p;

        public void solve(InputReader in, PrintWriter out){
            int n = in.nextInt();  

            p = new int[n];  

            edges = new int[n][];

            int[] from = new int[n-1];
            int[] to = new int[n-1];

            level = new int[n];
            marthaThere = new int[n];
            parents = new int[n];

            for(int i = 0;i<n;i++ ){
                level[i] = 0;
                marthaThere[i] = 0;
                parents[i] = -1;
            }   

            for(int i = 0;i<n-1;i++){
                from[i] = in.nextInt();
                from[i]--;
                to[i] = in.nextInt();
                to[i]--;
            }


            for (int i : from) p[i]++;
            for (int i : to) p[i]++;

            for (int i = 0; i < n; i++) edges[i] = new int[p[i]];

            for (int i = 0; i < from.length; i++) {
                edges[from[i]][--p[from[i]]] = to[i];
                edges[to[i]][--p[to[i]]] = from[i];
            }

            int x,y;
            x = in.nextInt();
            y = in.nextInt();
            if(x==y){
                out.println("0");
                return;
            }    
            x--;
            y--;
            martha = y;
            this.dfs(x,-1,0,0); 
            
            int val = (int)Math.ceil((double)(level[y] + 1 )/2.0);
            int ans = val;
            int temp = 0;
            for(int i =0;i<n;i++){
                //out.println(i+ " " + level[i] + " " + marthaThere[i]);
                if(level[i]==val && marthaThere[i]==1){
                    int var = this.dfs1(i,parents[i]);
                    temp = Math.max(temp,var);
                }
            }
            ans += temp;
            out.println(ans);
            return;
        }

        public void dfs(int node, int parent,int depth,int flag){
            level[node] = depth;
            marthaThere[node] = flag;
            parents[node] = parent;
            if(node==martha) {
                flag = 1;
                marthaThere[node] = 1;
            }
            for(int i : edges[node]){
                if(parent==i) continue;
                this.dfs(i,node,depth+1,flag);
                if(marthaThere[i]==1) marthaThere[node] = 1;
            }
            return;
        }

        public int dfs1(int node,int parent){
            if(edges[node].length==1) return 0;
            int ans = 0;
            for(int i : edges[node]){
                if(i==parent) continue;
                ans = Math.max(ans,dfs1(i,node));
            }
            return ans + 1;
        }

    }    
 
    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;
 
        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }
 
        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }
 
        public long nextLong() {
            return Long.parseLong(next());
        }
 
        public int nextInt() {
            return Integer.parseInt(next());
        }
 
    }
}